dev-lang/ghc/ghc-9.8.4-r1.ebuild

```
@@ -0,0 +1,942 @@
+inherit python-any-r1
+inherit autotools bash-completion-r1 flag-o-matic ghc-package
+inherit toolchain-funcs prefix check-reqs llvm unpacker haskell-cabal verify-sig
+
+DESCRIPTION="The Glasgow Haskell Compiler"
+HOMEPAGE="https://www.haskell.org/ghc/"
+
+VERIFY_SIG_OPENPGP_KEY_PATH=/usr/share/openpgp-keys/ghc.asc
+
+GHC_BRANCH_COMMIT="a3401159f2846605abb517e71af463df47398e72" # ghc-9.8.4-release
+
+GHC_BINARY_PV="9.6.2"
+SRC_URI="
+	https://downloads.haskell.org/~ghc/${PV}/${P}-src.tar.xz
+	verify-sig? ( https://downloads.haskell.org/~ghc/${PV}/${P}-src.tar.xz.sig )
+	!ghcbootstrap? (
+		https://downloads.haskell.org/~ghc/9.8.2/hadrian-bootstrap-sources/hadrian-bootstrap-sources-${GHC_BINARY_PV}.tar.gz
+		amd64? ( https://downloads.haskell.org/~ghc/${GHC_BINARY_PV}/ghc-${GHC_BINARY_PV}-x86_64-alpine3_12-linux-static-int_native.tar.xz )
+		arm64? ( elibc_glibc? (
+			https://downloads.haskell.org/~ghc/${GHC_BINARY_PV}/ghc-${GHC_BINARY_PV}-aarch64-deb10-linux.tar.xz
+		) )
+	)
+	test? (
+		https://gitlab.haskell.org/ghc/ghc/-/archive/${GHC_BRANCH_COMMIT}.tar.gz
+			-> ${PN}-${GHC_BRANCH_COMMIT}.tar.gz
+	)
+"
+
+GHC_BUGGY_TESTS=(
+	# Actual stderr output differs from expected:
+	# +/usr/libexec/gcc/x86_64-pc-linux-gnu/ld: warning: ManySections.o: missing .note.GNU-stack section implies executable stack
+	# +/usr/libexec/gcc/x86_64-pc-linux-gnu/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
+	"testsuite/tests/driver/recomp015"
+)
+
+yet_binary() {
+	case ${ARCH} in
+		amd64)
+			return 0
+			;;
+		arm64)
+			case "${ELIBC}" in
+				glibc)
+					return 0
+					;;
+				*)
+					return 1
+					;;
+			esac
+			;;
+		*)
+			return 1
+			;;
+	esac
+}
+
+# We are using the upstream static Alpine Linux binaries to bootstrap some
+# archs. These binaries have different properties than the ones we build
+# ourselves, so we need a way to check to see if they are in use.
+upstream_binary() {
+	case ${ARCH} in
+		amd64)
+			return 0
+			;;
+		arm64)
+			case "${ELIBC}" in
+				glibc)
+					return 0
+					;;
+				*)
+					return 1
+					;;
+			esac
+			;;
+		*)
+			return 1
+			;;
+	esac
+}
+
+# The location of the unpacked Alpine Linux tarball
+ghc_bin_path() {
+	local ghc_bin_triple
+	case ${ARCH} in
+		amd64)
+			ghc_bin_triple="x86_64-unknown-linux"
+			;;
+		arm64)
+			if use elibc_musl; then
+				ghc_bin_triple="aarch64-alpine-linux"
+			else
+				ghc_bin_triple="aarch64-unknown-linux"
+			fi
+			;;
+		*)
+			die "Unknown ghc binary triple. The list here should match yet_binary."
+			;;
+	esac
+
+	echo "${WORKDIR}/ghc-${GHC_BINARY_PV}-${ghc_bin_triple}"
+}
+
+GHC_PV=${PV}
+#GHC_PV=8.10.0.20200123 # uncomment only for -alpha, -beta, -rc ebuilds
+GHC_P=${PN}-${GHC_PV} # using ${P} is almost never correct
+
+S="${WORKDIR}"/${GHC_P}
+
+BUMP_LIBRARIES=(
+	# "hackage-name          hackage-version"
+	#Match 9.6.7
+	"directory 1.3.9.0"
+	"file-io 0.1.5"
+
+	#Match 9.12.2
+	"filepath 1.5.4.0"
+)
```

`os-string` is required by the new `filepath`.

```suggestion
BUMP_LIBRARIES=(
	# "hackage-name          hackage-version"
	#Match 9.6.7
	"directory 1.3.9.0"
	"file-io 0.1.5"

	#Match 9.12.2
	"os-string 2.0.7"
	"filepath 1.5.4.0"
)
```

---

This revision will need to be added to `CABAL_CORE_LIB_GHC_PV` in `os-string-2.0.7.ebuild` as well. (Note the special syntax that lets us specify a gentoo revision.)

```diff
-CABAL_CORE_LIB_GHC_PV="9.10.1 9.12.1 9.12.2"
+CABAL_CORE_LIB_GHC_PV="PM:9.8.4-r1 9.10.1 9.12.1 9.12.2"
```

---

I've been trying different methods, and I don't see a path forward for bumping to `filepath-1.5.4.0` in `ghc-9.8.4`. I tried using `hadrian` from `ghc-9.12.2`, but there are too many incompatibilities (such as `ghc-9.8.4` not having `ghc-platform` nor `ghc-toolchain`). It may be possible to build a static `hadrian-9.12.2` and include it, but `USE=ghcbootstrap` would be messy because of requirements on later `Cabal`.

`hadrian` from `ghc-9.6.7` is more compatible, but has no notion of `os-string`. However, it may be of use for bumping to `directory 1.3.9.0` in `ghc-9.8.4`.